apps/ewallet/lib/ewallet/gates/transactions/blockchain_local_gate.ex

```diff
@@ -0,0 +1,298 @@
+# Copyright 2018-2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule EWallet.TransactionGate.BlockchainLocal do
+  @moduledoc """
+  Handles the logic for a transaction of value from an account to a user. Delegates the
+  actual transaction to EWallet.LocalTransactionGate once the wallets have been loaded.
+  """
+  alias EWallet.TransactionFormatter
+  alias EWalletDB.{BlockchainTransactionState, Transaction, TransactionState, TransactionType}
+  alias EWalletDB.Helpers.Preloader
+  alias ActivityLogger.System
+  alias LocalLedger.Transaction, as: LedgerTransaction
+
+  @pending TransactionState.pending()
+  @confirmed TransactionState.confirmed()
+  @failed TransactionState.failed()
+  @blockchain_submitted TransactionState.blockchain_submitted()
+  @ledger_pending_blockchain_confirmed TransactionState.ledger_pending_blockchain_confirmed()
+  @blockchain_transaction_confirmed BlockchainTransactionState.confirmed()
+  @blockchain_transaction_failed BlockchainTransactionState.failed()
+
+  @default_error_code "blockchain_transaction_error"
+  @default_error_description "An error occured when processing this blockchain transaction"
+
+  # This is called by the transaction tracker when a blockchain transaction
+  # reaches the `confirmed` state or by the EWallet.TransactionGate.Blockchain module.
+  def process_with_transaction(transaction) do
+    transaction
+    |> TransactionType.get()
+    |> process_with_transaction(transaction)
+  end
+
+  # If the transaction is in a final state we don't need to proccess it
+  defp process_with_transaction(_, %{status: @confirmed} = transaction), do: {:ok, transaction}
+  defp process_with_transaction(_, %{status: @failed} = transaction), do: {:error, transaction}
+
+  defp process_with_transaction(
+         :from_blockchain_to_ewallet,
+         %{blockchain_transaction: %{status: @blockchain_transaction_confirmed}} = transaction
+       ) do
+    TransactionState.transition_to(
+      :from_blockchain_to_ewallet,
+      TransactionState.confirmed(),
+      transaction,
+      %{originator: %System{}}
+    )
+  end
+
+  defp process_with_transaction(
+         :from_blockchain_to_ewallet,
+         transaction
```

Comment:
By this logic it seems if by any chance `BlockchainTransactionState.submitted()` or `BlockchainTransactionState.pending_confirmations()` is passed, it will be considered a failed transaction. 

Maybe guard this with `%{blockchain_transaction: %{status: @blockchain_transaction_failed}}` so if those other statuses reach by mistake, they hang on to their current status rather than going to failed? Since a mistakenly failed transaction will be hard to recover from.

Also applicable to the rest of the module.